### PR TITLE
ArC - register synthetic beans for List injection points

### PR DESCRIPTION
**Pull request by mkouba**
- unless a matching bean exists

### CODE REVIEW

**Comment by mkouba** [comment]
Here's an idea - `javax.enterprise.inject.Instance` is not very intuitive when it comes to injection of multiple beans which is perfectly legal although not very common in the CDI world. In this PR, we register a synthetic bean for any `List<>` injection point for which a regular bean does not exist. As a result, a user can do e.g. `@Inject List<Service>` even if there's no such bean that would match this type (per type-safe resolution rules). There's a [test](https://github.com/quarkusio/quarkus/pull/21378/files#diff-38eda968310494cb1d339dc9f07cf3d098b76a947ef705d35e0e76691351f08b) that demonstrates the usage.

Improvement ideas:
- support `List<InstanceHandle<Service>>` as well
- the list could be sorted, e.g. the community discussed the possibility to use the `@Priority`

**Comment by mkouba** [comment]
CC @ebullient @manovotn @Ladicek @stuartwdouglas

**Comment by Ladicek** [comment]
I saw the discussion on Zulip and thought about it for a while. I think it's full of corner cases. The most obvious one: what happens when a bean of type `List<Service>` suddenly appears, when previously there was none? (Rhetorical question. I know what happens. I don't think it's good.) Note that our CDI container is flat, so adding a dependency often means adding new beans. Also note that there are libraries that actually do provide beans of collection types (I can't immediately think of anything actually providing beans of type `java.util.List`, but I recall the Quarkus Infinispan extension provides beans of type `Cache`, which transitively extends `java.util.Map`, so this happens easier than you might think).

I ended up thinking that this would be best done either using a dedicated type (`BeanCollection`? `InstanceCollection`?), or a dedicated annotation (`@InjectAll`?), otherwise it's too brittle.

**Comment by mkouba** [comment]
> The most obvious one: what happens when a bean of type List<Service> suddenly appears, when previously there was none? (Rhetorical question. I know what happens. I don't think it's good.) 

But this can happen even now. You add a dependency and get a new bean for type `Foo` and boom...

>  Note that our CDI container is flat, so adding a dependency often means adding new beans. Also note that there are libraries that actually do provide beans of collection types...

Well, the libraries should not provide beans of type `List<SomeApplicationClass>` which is the main use case. 

> I ended up thinking that this would be best done either using a dedicated type (BeanCollection? InstanceCollection?), or a dedicated annotation (@InjectAll?), otherwise it's too brittle.

I have a problem with those names ^. Yes, initially I was thinking about a dedicated type but that's not very convenient. I was also thinking about a special qualifier (e.g. `@InjecAll List<Service>` or even `@Inject @All List<Service>`) which would be the least evil but still not the most intuitive.

**Comment by mkouba** [comment]
Hm, I actually quite like the `@Inject @All List<Service>`...

**Comment by Ladicek** [comment]
> > The most obvious one: what happens when a bean of type List suddenly appears, when previously there was none? (Rhetorical question. I know what happens. I don't think it's good.)
> 
> But this can happen even now. You add a dependency and get a new bean for type `Foo` and boom...

Boom is good (especially in our case, when it fails the build). Silently changing what is injected, not so much.

> > Note that our CDI container is flat, so adding a dependency often means adding new beans. Also note that there are libraries that actually do provide beans of collection types...
> 
> Well, the libraries should not provide beans of type `List<SomeApplicationClass>` which is the main use case.

Fair point... until you consider parametric bean types :-) Here's the Infinispan example I mentioned above: https://github.com/quarkusio/quarkus/blob/2.4.1.Final/extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientProducer.java#L283

Haha. Yes. It's an advanced case. Sorry, I somehow got into a mode where this kind of crap pop up into my brain almost instantly :laughing:

**Comment by mkouba** [comment]
> > > The most obvious one: what happens when a bean of type List suddenly appears, when previously there was none? (Rhetorical question. I know what happens. I don't think it's good.)
> > 
> > 
> > But this can happen even now. You add a dependency and get a new bean for type `Foo` and boom...
> 
> Boom is good (especially in our case, when it fails the build). Silently changing what is injected, not so much.

It does not always fail the build. Think about `javax.enterprise.inject.Instance`.

> 
> > > Note that our CDI container is flat, so adding a dependency often means adding new beans. Also note that there are libraries that actually do provide beans of collection types...
> > 
> > 
> > Well, the libraries should not provide beans of type `List<SomeApplicationClass>` which is the main use case.
> 
> Fair point... until you consider parametric bean types :-) Here's the Infinispan example I mentioned above: https://github.com/quarkusio/quarkus/blob/2.4.1.Final/extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientProducer.java#L283

Parameterized bean types are OK. What's not OK is to define a bean with bean type `List<T>` or `Map<K,V>`. I'm not sure if it's the case of the infinispan producer. Maybe they should even use `@Typed(RemoteCache.class)`...

> 
> Haha. Yes. It's an advanced case. Sorry, I somehow got into a mode where this kind of crap pop up into my brain almost instantly laughing

No problem ;-). I will modify the PR and introduce the `@All` qualifier...

**Comment by Ladicek** [comment]
> > > > The most obvious one: what happens when a bean of type List suddenly appears, when previously there was none? (Rhetorical question. I know what happens. I don't think it's good.)
> > > 
> > > 
> > > But this can happen even now. You add a dependency and get a new bean for type `Foo` and boom...
> > 
> > 
> > Boom is good (especially in our case, when it fails the build). Silently changing what is injected, not so much.
> 
> It does not always fail the build. Think about `javax.enterprise.inject.Instance`.

Fair enough. But it would still fail at runtime with ambiguous dependency (where the `Instance` used to resolve to 1 bean, it would now "resolve" to 2). Or not, if the user doesn't call `get` but instead iterate over what the `Instance` has. At which point, presumably the author _expects_ that there may be multiple beans of the type, so hard to tell.

> > > > Note that our CDI container is flat, so adding a dependency often means adding new beans. Also note that there are libraries that actually do provide beans of collection types...
> > > 
> > > 
> > > Well, the libraries should not provide beans of type `List<SomeApplicationClass>` which is the main use case.
> > 
> > 
> > Fair point... until you consider parametric bean types :-) Here's the Infinispan example I mentioned above: https://github.com/quarkusio/quarkus/blob/2.4.1.Final/extensions/infinispan-client/runtime/src/main/java/io/quarkus/infinispan/client/runtime/InfinispanClientProducer.java#L283
> 
> Parameterized bean types are OK.

Parameterized (`List<String>`) yes. Parametric (`List<T>`) no. I don't have better terminology unfortunately. (JLS uses "generic" instead of "parametric", but I'm not sure if that would be an improvement.)

> What's not OK is to define a bean with bean type `List<T>` or `Map<K,V>`. I'm not sure if it's the case of the infinispan producer. Maybe they should even use `@Typed(RemoteCache.class)`...

Agree. My point is that it is rather easy to create a bean of type `List<T>` without even knowing.

> No problem ;-). I will modify the PR and introduce the `@All` qualifier...

I don't want to force you to do that, maybe other people will convince you that my objections are crap and should be ignored :-)

**Comment by stuartwdouglas** [comment]
IMHO it would be very unusual for there to be a conflict like we are talking about. 

Personally I am +1 to this change, as it makes things a lot more intuitive for users.

**Comment by geoand** [comment]
Definitely +1 from me. The use of the `@All` qualifier makes sense to me to disambiguate things.

**Comment by mkouba** [comment]
I've just pushed the `@Inject @All` solution. The implementation is a bit more complex but the UX is quite nice and we're more safe. In addition, the presence of `@All` is good for clarity - the javadoc defines the contract.

**Comment by ebullient** [comment]
This is great! But I don't know that it will change much for me as I'm using a recorder (to fetch/iterate instances programmatically) rather than using injection. Unless @geoand can prove me wrong, he's been fixing all of my early mistakes. ;)

**Comment by geoand** [comment]
> This is great! But I don't know that it will change much for me as I'm using a recorder (to fetch/iterate instances programmatically) rather than using injection. Unless @geoand can prove me wrong, he's been fixing all of my early mistakes. ;)

Which pieces do you have in mind? I'd be glad to take a look

**Comment by manovotn** [comment]
+1, this idea sounds good and adding `@All` is definitely a good call to distinguish it.

**Comment by ebullient** [comment]
> > This is great! But I don't know that it will change much for me as I'm using a recorder (to fetch/iterate instances programmatically) rather than using injection. Unless @geoand can prove me wrong, he's been fixing all of my early mistakes. ;)
> 
> Which pieces do you have in mind? I'd be glad to take a look

Brain dumping to poor @geoand in DMs. Have pity on his 🧠

**Comment by geoand** [comment]
> Brain dumping to poor @geoand in DMs. Have pity on his brain

It's all good! :)

**Comment by mkouba** [comment]
I think that this PR is ready for review now ;-)

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java [hunk 1]
```diff
@@ -612,6 +643,190 @@ BeanDefiningAnnotationBuildItem quarkusMain() {
...
+                continue;
+            }
+
+            BeanConfigurator<?> configurator = beanRegistrationPhase.getContext()
+                    .configure(List.class)
+                    .forceApplicationClass()
+                    .scope(BuiltinScope.DEPENDENT.getInfo())
+                    .addType(injectionPoint.getRequiredType());
+
+            injectionPoint.getRequiredQualifiers().forEach(configurator::addQualifier);
```
Comment (Ladicek) [inline extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java:760]: Just wondering if this should read the `qualifiers` set created above, instead of `injectionPoint.getRequiredQualifiers()`?

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Identified.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import static java.lang.annotation.ElementType.TYPE;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+import javax.enterprise.util.AnnotationLiteral;
+import javax.inject.Qualifier;
+
+/**
+ * Qualify a bean with a string-based identifier.
```
Comment (Ladicek) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Identified.java:15]: Maybe drop this and use https://github.com/smallrye/smallrye-common/blob/main/annotation/src/main/java/io/smallrye/common/annotation/Identifier.java?

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java [hunk 3]
```diff
@@ -612,6 +643,190 @@ BeanDefiningAnnotationBuildItem quarkusMain() {
...
+
+            @Override
+            public boolean appliesTo(Kind kind) {
+                return kind == Kind.FIELD || kind == Kind.METHOD;
+            }
+
+            @Override
+            public void transform(TransformationContext ctx) {
+                if (Annotations.contains(ctx.getAnnotations(), DotNames.ALL)) {
+                    // For each injection point annotated with @All add a synthetic qualifier
```
Comment (Ladicek) [inline extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java:663]: Out of curiosity, why is this necessary?

**Review by Ladicek — APPROVED** [review]
Can't say I really understand it, but it seems fine :-)

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java (hunk 1 above)
Comment (mkouba) [reply extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java:760]: I don't think so. The synthetic bean must copy the qualifiers on the injection point, e.g. `@All`, whereas the `qualifiers` created above are used for dynamic lookup when the list is created at runtime, e.g. `@Any`.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Identified.java (hunk 2 above)
Comment (mkouba) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Identified.java:15]: Yeah, it's the same semantics but we don't want to add a dependency on `smallrye-common`.

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java (hunk 3 above)
Comment (mkouba) [reply extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java:663]: That's  a good question. Imagine we have the following injection points:

```java
   @Inject
   @All
   List<Converter> converters;

   @Inject
   @All
   @MyQualifier
   List<Converter> convertersMyQualifier;
```
We need to generate two synthetic beans here. One for `converters` to perform a dynamic lookup identical to `@Inject @Any Instance<Converter>` and one for `convertersMyQualifier` to perfom a dynamic lookup identical to `@Inject @MyQualifier Instance<Converter>`. The first one must satisfy injection points with required type `List<Converter>` and qualifier `@All`. The other one must satisfy injection points with required type `List<Converter>` and qualifiers `@All` and  `@MyQualifier`. The problem is that if we don't add a synthetic qualifier `@Identified("foo")` (where foo is a hash of required type and required qualifiers) then the second bean would also satisfy the first injection point - the required type matches and the bean has qualifier `@All`. Does it make sense?

Comment (Ladicek) [reply extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java:663]: Aaaaaha! Yes it does, I suspected there's a good reason for it, I just couldn't see it.

I guess it's because `@All` is a regular qualifier -- if it was a special annotation, things would probably look a lot different.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Identified.java (hunk 2 above)
Comment (Ladicek) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Identified.java:15]: I'm just thinking we probably shouldn't have 2 annotations with the same meaning available as part of public API -- and IIUC, this `@Identified` is meant to be public API?

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java (hunk 1 above)
Comment (Ladicek) [reply extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java:760]: Gotcha, thanks!

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java (hunk 3 above)
Comment (mkouba) [reply extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java:663]: > I guess it's because `@All` is a regular qualifier -- if it was a special annotation, things would probably look a lot different.

Well, it needs to be a qualifier because it's a regular injection point. If it wasn't then `@Inject @All List<T>` would be satisfied with a any bean that has `List<T>` in its set of bean types and qualifier `@Default`.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Identified.java (hunk 2 above)
Comment (mkouba) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Identified.java:15]: > this @Identified is meant to be public API

Yes, the `io.quarkus.arc` package is meant to be public API. Ok, let me try to add the `smallrye-common-annotation` dependency.

Comment (mkouba) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Identified.java:15]: Actually, we shouldn't use a public API annotation at all. It should be an internal qualifier instead. Because otherwise we could run into the same problem - a collision with user-provided qualifier.

Comment (mkouba) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Identified.java]: I've moved the qualifier annotation into the `io.quarkus.arc.impl` package.

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java (hunk 3 above)
Comment (Ladicek) [reply extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java:663]: I'm pretty sure we could define `@All` to be a special annotation that affects how `@Inject` works. Qualifier works too, for sure.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Identified.java (hunk 2 above)
Comment (Ladicek) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Identified.java]: Ah good point.

File: extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java (hunk 3 above)
Comment (Ladicek) [reply extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java:663]: > I'm pretty sure we could define `@All` to be a special annotation that affects how `@Inject` works.

Actually that's probably why I originally suggested `@InjectAll` :-) But as I said, standalone `@All` is fine too.

Comment (mkouba) [reply extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java:663]: We could but I don't think it's a good idea to override the type-safe resolution rules. It could be even more confusing :-).

**Review by gastaldi — APPROVED** [review]
Good idea!

**Comment by gsmet** [comment]
This is too big to be backported post CR1.

**Comment by mkouba** [comment]
> This is too big to be backported post CR1.

No problem.
